libs/wingc/src/type_check.rs

```diff
@@ -2922,4 +2951,126 @@ mod tests {
 		assert!(!Phase::Preflight.is_subtype_of(&Phase::Inflight));
 		assert!(!Phase::Inflight.is_subtype_of(&Phase::Preflight));
 	}
+
+	#[test]
+	fn optional_subtyping() {
+		let string = UnsafeRef::<Type>(&Type::String as *const Type);
+		let opt_string = UnsafeRef::<Type>(&Type::Optional(string) as *const Type);
+
+		// T is a subtype of T? since T can be used anywhere a T? is expected
+		// (but not vice versa)
+		assert!(string.is_subtype_of(&opt_string));
+		assert!(!opt_string.is_subtype_of(&string));
+
+		// subtyping is reflexive
+		assert!(string.is_subtype_of(&string));
+		assert!(opt_string.is_subtype_of(&opt_string));
+	}
+
+	#[test]
+	fn function_subtyping_across_phases() {
+		let void = UnsafeRef::<Type>(&Type::Void as *const Type);
+		let inflight_fn = Type::Function(FunctionSignature {
+			parameters: vec![],
+			return_type: void,
+			phase: Phase::Inflight,
+			js_override: None,
+		});
+		let preflight_fn = Type::Function(FunctionSignature {
+			parameters: vec![],
+			return_type: void,
+			phase: Phase::Preflight,
+			js_override: None,
+		});
+
+		// functions of different phases are not subtypes of each other
+		assert!(!inflight_fn.is_subtype_of(&preflight_fn));
+		assert!(!preflight_fn.is_subtype_of(&inflight_fn));
+
+		// subtyping is reflexive
+		assert!(inflight_fn.is_subtype_of(&inflight_fn));
+		assert!(preflight_fn.is_subtype_of(&preflight_fn));
+	}
+
+	#[test]
+	fn function_subtyping_incompatible_single_param() {
+		let void = UnsafeRef::<Type>(&Type::Void as *const Type);
+		let num = UnsafeRef::<Type>(&Type::Number as *const Type);
+		let string = UnsafeRef::<Type>(&Type::String as *const Type);
+		let num_fn = Type::Function(FunctionSignature {
+			parameters: vec![num],
+			return_type: void,
+			phase: Phase::Inflight,
+			js_override: None,
+		});
+		let str_fn = Type::Function(FunctionSignature {
+			parameters: vec![string],
+			return_type: void,
+			phase: Phase::Inflight,
+			js_override: None,
+		});
```

Comment:
something like this might be clearer (also for all the other tests)
```suggestion
        let sig = FunctionSignature {
			parameters: vec![num],
			return_type: void,
			phase: Phase::Inflight,
			js_override: None,
		};
		let num_fn = Type::Function(sig);
		let str_fn = Type::Function(FunctionSignature {
			parameters: vec![string], ..sig.clone()
		});
```